### PR TITLE
Add typehints to result_set

### PR DESCRIPTION
**Pull request by souravsingh**
A work in progress PR to add type hints to result_set.py

### CODE REVIEW

**Comment by drvinceknight** [comment]
The tests are failing because of syntax errors.

**Comment by drvinceknight** [comment]
83838f5 another error that comes up on import. On http://axelrod.readthedocs.io/en/stable/tutorials/contributing/running_tests.html there is info about running a specific test file instead of running the whole test suite if that's helpful (this would at least save you time of pushing code that doesn't run).

**Comment by drvinceknight** [comment]
Also can you add this file to https://github.com/Axelrod-Python/Axelrod/blob/master/run_mypy.py

File: run_mypy.py [hunk 1]
```diff
@@ -11,6 +11,7 @@
            "axelrod/moran.py",
            "axelrod/plot.py",
            "axelrod/random_.py",
+           "axelrod/result_set.py"
```
Comment (drvinceknight) [inline run_mypy.py]: You are missing a `,` here.

Information about running the type check tests is available here: http://axelrod.readthedocs.io/en/stable/tutorials/contributing/running_tests.html#type-checking

**Comment by drvinceknight** [comment]
Quite a few type checking errors:

```
76.09s$ python run_mypy.py
axelrod/result_set.py:409: error: Need type annotation for variable
axelrod/result_set.py:410: error: Need type annotation for variable
axelrod/result_set.py:414: error: Need type annotation for variable
axelrod/result_set.py:417: error: Need type annotation for variable
axelrod/result_set.py:419: error: Need type annotation for variable
axelrod/result_set.py:458: error: Cannot determine type of 'payoffs'
axelrod/result_set.py:460: error: Cannot determine type of 'payoffs'
axelrod/result_set.py:492: error: Cannot determine type of 'normalised_cooperation'
axelrod/result_set.py:493: error: Cannot determine type of 'normalised_cooperation'
axelrod/result_set.py:551: error: Cannot determine type of 'normalised_scores'
axelrod/result_set.py:598: error: Cannot determine type of 'state_distribution'
axelrod/result_set.py:601: error: Cannot determine type of 'state_distribution'
axelrod/result_set.py:616: error: Cannot determine type of 'state_to_action_distribution'
axelrod/result_set.py:621: error: Cannot determine type of 'state_to_action_distribution'
axelrod/result_set.py:919: error: "bool" has no attribute "update"
axelrod/result_set.py:923: error: "bool" has no attribute "close"
axelrod/result_set.py:938: error: Need type annotation for variable
axelrod/result_set.py:939: error: Need type annotation for variable
axelrod/result_set.py:945: error: "bool" has no attribute "update"
axelrod/result_set.py:948: error: "bool" has no attribute "close"
axelrod/result_set.py:1040: error: Argument 1 to "_update_repetitions" of "ResultSet" has incompatible type "Tuple[int, int]"; expected "Tuple[Any]"
axelrod/result_set.py:1041: error: Argument 1 to "_update_players" of "ResultSet" has incompatible type "Tuple[int, int]"; expected "Tuple[Any]"
axelrod/result_set.py:1043: error: "bool" has no attribute "update"
axelrod/result_set.py:1046: error: "bool" has no attribute "close"
axelrod/result_set.py:1084: error: "bool" has no attribute "update"
axelrod/result_set.py:1091: error: "bool" has no attribute "close"
```

Information about running the type checking testing locally is available here: http://axelrod.readthedocs.io/en/stable/tutorials/contributing/running_tests.html#type-checking

**Comment by souravsingh** [comment]
mypy is not working. This is the problem-
```shell
No command 'mypy' found, did you mean:
 Command 'mpy' from package 'yorick-mpy-mpich2' (universe)
 Command 'mpy' from package 'yorick-mpy-openmpi' (universe)
 Command 'pypy' from package 'pypy' (universe)
mypy: command not found
```

I did a ```pip freeze``` to check installation and it is present-
```
mypy==0.511
mypy-lang==0.4.6
```

**Comment by drvinceknight** [comment]
> mypy is not working. This is the problem-

It's difficult for me to help with that I'm afraid. Perhaps you're not running your mypy command in the same environment? FYI,`pip show mypy` will display the currently installed version of mypy.

**Comment by drvinceknight** [comment]
`pip show mypy` will also tell you the path to the file so potentially you could call the script with the fully path. Here's my output for example:

```
Axelrod(master): pip show mypy
---
Metadata-Version: 2.0
Name: mypy
Version: 0.501
Summary: Optional static typing for Python
Home-page: http://www.mypy-lang.org/
Author: Jukka Lehtosalo
Author-email: jukka.lehtosalo@iki.fi
Installer: pip
License: MIT License
Location: /home/vince/anaconda3/lib/python3.5/site-packages
Requires: typed-ast
...
```

See the Location field, which tells me that my `mypy` install is in the root anaconda install, as that's in my path I can call `mypy` directly or I can run:

```
$ /home/vince/anaconda3/bin/mypy -h
```

I can also use `which` to find out where the `mypy` package is that is recognised (this wouldn't help in your case as it doesn't look like `mypy` is in your path).

```
Axelrod(master): which mypy
/home/vince/anaconda3/bin/mypy
```

**Comment by souravsingh** [comment]
I will be resending the PR soonish

**Pull request closed without merging**